src/OpenSpaceToolkit/Astrodynamics/Dynamics/Thruster.cpp

```diff
@@ -12,29 +19,98 @@ namespace astro
 namespace dynamics
 {
 
-Thruster::Thruster(const SatelliteSystem& aSatelliteSystem, const String& aName)
-    : name_(aName),
-      satelliteSystem_(aSatelliteSystem)
+using ostk::math::obj::VectorXd;
+using ostk::math::obj::Vector3d;
+
+using ostk::physics::units::Mass;
+
+using ostk::astro::trajectory::state::CoordinatesSubset;
+using ostk::astro::trajectory::state::coordinatessubsets::CartesianPosition;
+using ostk::astro::trajectory::state::coordinatessubsets::CartesianVelocity;
+
+Thruster::Thruster(
+    const SatelliteSystem& aSatelliteSystem, const Shared<const GuidanceLaw>& aGuidanceLaw, const String& aName
+)
+    : Dynamics(aName),
+      satelliteSystem_(aSatelliteSystem),
+      guidanceLaw_(aGuidanceLaw)
 {
 }
 
 Thruster::~Thruster() {}
 
-String Thruster::getName() const
+SatelliteSystem Thruster::getSatelliteSystem() const
+{
+    if (!satelliteSystem_.isDefined())
+    {
+        throw ostk::core::error::runtime::Undefined("Satellite System");
+    }
+
+    return satelliteSystem_;
+}
+
+Shared<const GuidanceLaw> Thruster::getGuidanceLaw() const
 {
-    return name_;
+    return guidanceLaw_;
 }
 
-SatelliteSystem Thruster::getSatelliteSystem() const
+Array<Shared<const CoordinatesSubset>> Thruster::getReadCoordinatesSubsets() const
 {
-    return satelliteSystem_;
+    return {
+        CartesianPosition::Default(),
+        CartesianVelocity::Default(),
+        CoordinatesSubset::Mass(),
+    };
+}
+
+Array<Shared<const CoordinatesSubset>> Thruster::getWriteCoordinatesSubsets() const
+{
+    return {
+        CartesianVelocity::Default(),
+        CoordinatesSubset::Mass(),
+    };
+}
+
+bool Thruster::isDefined() const
+{
+    return satelliteSystem_.isDefined();
+}
+
+VectorXd Thruster::computeContribution(
+    const Instant& anInstant, const VectorXd& x, [[maybe_unused]] const Shared<const Frame>& aFrameSPtr
+) const
+{
+    const Vector3d positionCoordinates = {x[0], x[1], x[2]};
+    const Vector3d velocityCoordinates = {x[3], x[4], x[5]};
+
+    if (x[6] <= satelliteSystem_.getMass().inKilograms())  // We compare against the dry mass of the Satellite
+    {
+        throw ostk::core::error::RuntimeError("Out of fuel.");
+    }
+
+    const Real thrustAccelerationMagnitude =
+        satelliteSystem_.accessPropulsionSystem().getAcceleration(Mass::Kilograms(x[6])).getValue();
+
+    const Vector3d acceleration = guidanceLaw_->computeAcceleration(
+        anInstant, positionCoordinates, velocityCoordinates, thrustAccelerationMagnitude, aFrameSPtr
+    );
```

Comment:
This is the key change IMO right? The isolation of the `acceleration` (a.k.a. the control input) computation based on another object plugged in here. This is where we could get `guidanceLaw_->computeAcceleration` to return a tabulated set of acceleration, do you agree? or do you see this at another level?

Reply:
Precisely. We can create a tabulated acceleration guidance law that just interpolates and returns an acceleration at the requested time.

Reply:
Aligned! I can tackle that step for Tabulated acceleration soon if you want ;)